commit e3867eb1ce2282f9ac4cc729a3b1beb74bb017d6
Author: Miroslav Lichvar <mlichvar@redhat.com>
Date:   Tue Nov 26 15:10:34 2024 +0100

    phc2sys: Don't require -O option without -a and -w.
    
    Allow phc2sys to be started in the static mode without the -O option,
    meaning it should fetch the UTC offset from ptp4l, even if the -w option
    is not included on the command line to wait for ptp4l to have a port in
    the master or slave state.
    
    This is useful when the PHC is synchronized by ts2phc or other
    application and the system clock needs to be synchronized to the PHC
    even when ptp4l has all ports in the faulty state (e.g. all network
    links are down).
    
    It's the responsibility of the user to start phc2sys when ptp4l is
    already running. phc2sys will exit with an error message if it cannot
    get a response with the 1-second timeout.
    
    Signed-off-by: Miroslav Lichvar <mlichvar@redhat.com>

diff --git a/phc2sys.8 b/phc2sys.8
index df0608a..5ca467d 100644
--- a/phc2sys.8
+++ b/phc2sys.8
@@ -161,7 +161,10 @@ minimize the error caused by random delays in scheduling and bus utilization.
 The default is 5.
 .TP
 .BI \-O " offset"
-Specify the offset between the sink and source times in seconds. Not
+Specify the offset between the sink and source times in seconds. If not
+specified, the offset will be set according to the currentUtcOffset value
+obtained from ptp4l and the direction of synchronization, assuming the system
+clock is keeping time in UTC and all used PHCs keep time in TAI. Not
 compatible with the
 .B \-a
 option.  See
diff --git a/phc2sys.c b/phc2sys.c
index 1745558..d8b0991 100644
--- a/phc2sys.c
+++ b/phc2sys.c
@@ -1406,12 +1406,6 @@ int main(int argc, char *argv[])
 		goto bad_usage;
 	}
 
-	if (!autocfg && !wait_sync && !settings.forced_sync_offset) {
-		fprintf(stderr,
-			"time offset must be specified using -w or -O\n");
-		goto bad_usage;
-	}
-
 	if (hardpps_configured(pps_fd) && (dst_cnt != 1 ||
 	    strcmp(dst_names[0], "CLOCK_REALTIME"))) {
 		fprintf(stderr,
@@ -1501,7 +1495,7 @@ int main(int argc, char *argv[])
 
 	r = -1;
 
-	if (wait_sync) {
+	if (wait_sync || !domains[0].forced_sync_offset) {
 		snprintf(uds_local, sizeof(uds_local),
 			 "/var/run/phc2sys.%d", getpid());
 
@@ -1516,7 +1510,7 @@ int main(int argc, char *argv[])
 				  phc2sys_recv_subscribed, &domains[0]))
 			goto end;
 
-		while (is_running()) {
+		while (wait_sync && is_running()) {
 			r = run_pmc_wait_sync(domains[0].agent, 1000);
 			if (r < 0)
 				goto end;
